src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs

```diff
@@ -91,26 +91,81 @@ internal partial class CSharpIntroduceVariableService
+                declarationStatement, isEntireLambdaBodySelected, rewrittenBody, doesDelegateMethodReturnVoid);
+            var newLambda = oldLambda.WithBody(newBody);
+
+            var newRoot = document.Root.ReplaceNode(oldLambda, newLambda);
+            return document.Document.WithSyntaxRoot(newRoot);
+        }
+
+        private static BlockSyntax GetNewBlockBodyForLambda(
+            LocalDeclarationStatementSyntax declarationStatement,
+            bool isEntireLambdaBodySelected,
+            ExpressionSyntax rewrittenBody,
+            bool doesDelegateMethodReturnVoid)
+        {
+            // For lambdas with void return types, we don't need to include the rewritten body if the entire lambda body
+            // was originally selected for refactoring, as the rewritten body should already be encompassed within the
+            // declaration statement.
+
+            // Case 1a: The lambda has a void return type, and the user selects the entire lambda body.
+            // e.g.:
+            //     Action<int> goo = x => [|x.ToString()|];
+            //
+            // After refactoring:
+            //     Action<int> goo = x =>
+            //     {
+            //         string v = x.ToString();
+            //     };
+            var newBody = SyntaxFactory.Block(declarationStatement);
+
+            if (doesDelegateMethodReturnVoid && !isEntireLambdaBodySelected)
+            {
+                // Case 1b: The user didn't select the entire lambda body.
+                // e.g.:
+                //     Task.Run(() => File.Copy("src", [|Path.Combine("dir", "file")|]));
+                //
+                // After refactoring:
+                //     Task.Run(() =>
+                //     {
+                //         string destFileName = Path.Combine("dir", "file");
+                //         File.Copy("src", destFileName);
+                //     });
+                newBody = newBody.AddStatements(
+                    SyntaxFactory.ExpressionStatement(rewrittenBody, SyntaxFactory.Token(SyntaxKind.SemicolonToken)));
+            }
+            else if (!doesDelegateMethodReturnVoid)
+            {
+                // Case 2: The lambda has a non-void return type.
+                // e.g.:
+                //     Func<int, int> f = x => [|x + 1|];
+                //
+                // After refactoring:
+                //     Func<int, int> f = x =>
+                //     {
+                //         var v = x + 1;
+                //         return v;
+                //     };
+                newBody = newBody.AddStatements(
+                    SyntaxFactory.ReturnStatement(rewrittenBody));
+            }
 
             // Add an elastic newline so that the formatter will place this new lambda body across multiple lines.
             newBody = newBody.WithOpenBraceToken(newBody.OpenBraceToken.WithAppendedTrailingTrivia(SyntaxFactory.ElasticCarriageReturnLineFeed))
```

Comment:
so, reading hte above, i think it would just be much simpler as:

```c#
if (doesDelegateMethodReturnVoid )
{
     if (isEntireLambdaBodySelected)
         return SyntaxFactory.Block(declarationStatement);
 
     return ...
}
else 
{
     return ...
}
```

Just have the caller fixup the braces.
}